### PR TITLE
UI Components Practical

### PR DESCRIPTION
**Pull request by jay-a-shah**

### CODE REVIEW

File: Demo app/Demo app/LoginAndSignUpScreen/WebViewController.swift [hunk 1]
```diff
@@ -9,18 +9,28 @@ import UIKit
 import WebKit
 
 class WebViewController: UIViewController {
-    
-    //MARK: - Outlet
+//MARK: - Outlet
```
Comment (purvesh-d) [inline Demo app/Demo app/LoginAndSignUpScreen/WebViewController.swift]: Why you have did this changes?

File: Demo app/Demo app/LoginAndSignUpScreen/WebViewController.swift [hunk 2]
```diff
@@ -9,18 +9,28 @@ import UIKit
...
     @IBOutlet weak var webView: WKWebView!
     @IBOutlet var searchBar: UISearchBar!
     
-    //MARK: -View Lifecycle
+//MARK: -View Lifecycle
+    
     override func viewDidLoad() {
         super.viewDidLoad()
-        guard let url = URL(string: Constants.googleUrl) else {
```
Comment (purvesh-d) [inline Demo app/Demo app/LoginAndSignUpScreen/WebViewController.swift]: Why you have did this changes?

File: Demo app/Demo app/LoginAndSignUpScreen/WebViewController.swift [hunk 3]
```diff
@@ -9,18 +9,28 @@ import UIKit
...
             return
         }
         let urlRequest = URLRequest(url: url)
         webView.load(urlRequest)
     }
 }
+
+//extension WebViewController: UISearchBarDelegate{
+//    func searchBarCancelButtonClicked(_ searchBar: UISearchBar) {
+//        searchBar.text = ""
```
Comment (purvesh-d) [inline Demo app/Demo app/LoginAndSignUpScreen/WebViewController.swift]: commented code

**Review by purvesh-d — CHANGES_REQUESTED** [review]

File: Demo app/Demo app/DatePickerView.swift [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+//MARK: OpenDatePicker
+extension DatePickerView{
+    
+    func openDatePicker(){
+        let datePicker = UIDatePicker()
+        datePicker.datePickerMode = .date
+        datePicker.preferredDatePickerStyle = .wheels
+        textViewSelectDate.inputView = datePicker
+        let toolBar = UIToolbar(frame: CGRect(x: 0, y: 0, width: self.view.frame.width, height: 44))
```
Comment (purvesh-d) [inline Demo app/Demo app/DatePickerView.swift]: move height to constant file

File: Demo app/Demo app/LoginPage.swift [hunk 5]
```diff
@@ -8,11 +8,18 @@
 import UIKit
 
 class LoginPage: UIViewController{
-    
-    //MARK: - View LifeCycle
+//MARK: - Outlets
+    @IBOutlet weak var passwordTextField: UITextField!
```
Comment (purvesh-d) [inline Demo app/Demo app/LoginPage.swift:13]: MARK indentation is not proper

File: Demo app/Demo app/LoginPage.swift [hunk 6]
```diff
@@ -8,11 +8,18 @@
...
+    @IBOutlet weak var emailTextField: UITextField!
     override func viewDidLoad() {
         super.viewDidLoad()
-        title = Constants.titleLoginPage
+        // Do any additional setup after loading the view.
```
Comment (purvesh-d) [inline Demo app/Demo app/LoginPage.swift]: remove unnecessary comments

File: Demo app/Demo app/LoginPage.swift [hunk 7]
```diff
@@ -29,6 +36,21 @@ extension LoginPage{
             view.backgroundColor = .purple
         }else if (sender.selectedSegmentIndex == Constants.FOUR){
             view.backgroundColor = .orange
+    }
+    
+}
```
Comment (purvesh-d) [inline Demo app/Demo app/LoginPage.swift]: code formating is not proper

File: Demo app/Demo app/LoginPage.swift [hunk 8]
```diff
@@ -29,6 +36,21 @@ extension LoginPage{
...
+}
+extension LoginPage: UITextFieldDelegate{
+    func textFieldShouldReturn(_ textField: UITextField) -> Bool {
+        switch textField {
+        case emailTextField:
+            //emailTextField.resignFirstResponder()
```
Comment (purvesh-d) [inline Demo app/Demo app/LoginPage.swift]: why commented code?

File: Demo app/Demo app/TabBarController/HomeViewController.swift [hunk 9]
```diff
@@ -10,14 +10,122 @@ import UIKit
...
+    func numberOfSections(in tableView: UITableView) -> Int {
+        return Constants.ONE
+    }
+   
+    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+        filteredData.count
+    }
+    
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        let cell = tableView.dequeueReusableCell(withIdentifier: "cell", for:indexPath)
```
Comment (purvesh-d) [inline Demo app/Demo app/TabBarController/HomeViewController.swift]: move identifier in constant file

File: Demo app/Demo app/TabBarController/HomeViewController.swift [hunk 10]
```diff
@@ -10,14 +10,122 @@ import UIKit
...
+}
+//MARK: - UICollectionViewDataSource
+extension HomeViewController: UICollectionViewDataSource{
+    
+    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
+        return addItems
+    }
+    
+    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
+        if let cell = collectionView.dequeueReusableCell(withReuseIdentifier: "cell", for: indexPath) as? UICollectionViewCell {
```
Comment (purvesh-d) [inline Demo app/Demo app/TabBarController/HomeViewController.swift]: move cell in constant file

File: Demo app/Demo app/TabBarController/MapViewController.swift [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+//MARK: - UIPickerViewDelegate
+extension MapViewController: UIPickerViewDelegate{
+    
+    func pickerView(_ pickerView: UIPickerView, titleForRow row: Int, forComponent component: Int) -> String? {
+        return data[row]
+    }
+    
+    func pickerView(_ pickerView: UIPickerView, didSelectRow row: Int, inComponent component: Int) {
+        switch row {
+        case Constants.ONE: redirectToLocation(row: row, latitude: Constants.gandinagarLatitude, longitude: Constants.gandinagarLongitude)
```
Comment (purvesh-d) [inline Demo app/Demo app/TabBarController/MapViewController.swift]: code formatting required

**Review by purvesh-d — APPROVED** [review]

File: Demo app/Demo app/DatePickerView.swift (hunk 4 above)
Comment (purvesh-d) [reply Demo app/Demo app/DatePickerView.swift]: pending

File: Demo app/Demo app/DatePickerView.swift [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+        let btnCancel = UIBarButtonItem(title: Constants.btnCancel, style: .plain , target: self, action: #selector(self.onClickCancelBtn))
+        let btnDone = UIBarButtonItem(title: Constants.btnDone, style: .plain , target: self, action: #selector(self.onClickDoneBtn))
+        toolBar.setItems([btnCancel,btnDone], animated: true)
+        textViewSelectDate.inputAccessoryView = toolBar
+    }
+   
+    @objc func onClickCancelBtn(){
```
Comment (purvesh-d) [inline Demo app/Demo app/DatePickerView.swift:49]: make separate extension for objc func

File: Demo app/Demo app/LoginPage.swift (hunk 5 above)
Comment (purvesh-d) [reply Demo app/Demo app/LoginPage.swift:13]: pending

**Review by purvesh-d — CHANGES_REQUESTED** [review]

File: Demo app/Demo app/DatePickerView.swift (hunk 4 above)
Comment (jay-a-shah) [reply Demo app/Demo app/DatePickerView.swift]: Done

File: Demo app/Demo app/ViewController.swift [hunk 13]
```diff
@@ -33,8 +33,8 @@ extension ViewController {
     @IBAction func goToSecondView(_ sender: UIButton) {
         if let controller = storyboard?.instantiateViewController(withIdentifier: Constants.secondViewController) as? SecondViewController {
             controller.delegate = self
-            self.present(controller, animated: true, completion: nil)
-        }
+            self.navigationController?.pushViewController(controller, animated: true)
+            }
```
Comment (purvesh-d) [inline Demo app/Demo app/ViewController.swift]: code formatting

**Review by purvesh-d — CHANGES_REQUESTED** [review]

File: Demo app/Demo app/TabBarController/ImageViewController.swift [hunk 14]
```diff
@@ -8,14 +8,19 @@
...
+    
     //MARK: - View LifeCycle
     override func viewDidLoad() {
         super.viewDidLoad()
-        imageView.layer.cornerRadius = 75
+        imageView.layer.cornerRadius = CGFloat(Constants.SEVENTY)
+        btnAddPhoto.layer.masksToBounds = false
+        btnAddPhoto.layer.shadowRadius = CGFloat(Constants.THIRTY)
+        btnAddPhoto.layer.shadowColor = UIColor.black.cgColor
+        btnAddPhoto.layer.shadowOffset = CGSize(width: 200, height: 200)
```
Comment (purvesh-d) [inline Demo app/Demo app/TabBarController/ImageViewController.swift]: constant width height

**Review by purvesh-d — CHANGES_REQUESTED** [review]

File: Demo app/Demo app/TabBarController/HomeViewController.swift [hunk 15]
```diff
@@ -10,14 +10,125 @@ import UIKit
...
+
+//MARK: - UISearchBarDelegate
+extension HomeViewController: UISearchBarDelegate{
+    
+    func searchBar(_ searchBar: UISearchBar, textDidChange searchText: String) {
+        filteredData = data
+        if searchText != "" {
+            filteredData = filteredData.filter{ $0.lowercased().contains(searchText.lowercased())}
+            tableView.reloadData()
+        }else{
```
Comment (purvesh-d) [inline Demo app/Demo app/TabBarController/HomeViewController.swift]: code formatting required

File: Demo app/Demo app/TabBarController/HomeViewController.swift [hunk 16]
```diff
@@ -10,14 +10,125 @@ import UIKit
...
+    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
+        if let cell = collectionView.dequeueReusableCell(withReuseIdentifier: Identifiers.tableViewTabBarCell.rawValue, for: indexPath) as? UICollectionViewCell {
+            return cell
+        }
+        return UICollectionViewCell()
+    }
+}
+
+//MARK: - UICollectionViewDelegate
+extension HomeViewController: UICollectionViewDelegate {
```
Comment (purvesh-d) [inline Demo app/Demo app/TabBarController/HomeViewController.swift]: remove this block if not required

**Review by purvesh-d — CHANGES_REQUESTED** [review]

**Review by purvesh-d — APPROVED** [review]

**Pull request merged**